pkg/transport/manager.go

```diff
@@ -74,49 +75,69 @@ func (tm *Manager) Serve(ctx context.Context) {
 	})
 }
 
-func (tm *Manager) serve(ctx context.Context) {
-	var listeners []*snet.Listener
+func (tm *Manager) serveNetwork(ctx context.Context, netType string) {
+	if tm.isClosing() {
+		return
+	}
 
-	for _, netType := range tm.n.TransportNetworks() {
-		if tm.isClosing() {
-			return
-		}
+	// this func may be called by either initiating routing or a callback,
+	// so we should check whether this type of network is already being served
+	tm.servingNetsMu.Lock()
+	if _, ok := tm.servingNets[netType]; ok {
+		tm.servingNetsMu.Unlock()
+		return
+	}
+	tm.servingNets[netType] = struct{}{}
+	tm.servingNetsMu.Unlock()
 
-		lis, err := tm.n.Listen(netType, skyenv.DmsgTransportPort)
-		if err != nil {
-			tm.Logger.WithError(err).Fatalf("failed to listen on network '%s' of port '%d'",
-				netType, skyenv.DmsgTransportPort)
-			continue
-		}
-		tm.Logger.Infof("listening on network: %s", netType)
-		listeners = append(listeners, lis)
+	lis, err := tm.n.Listen(netType, skyenv.DmsgTransportPort)
+	if err != nil {
+		tm.Logger.WithError(err).Fatalf("failed to listen on network '%s' of port '%d'",
+			netType, skyenv.DmsgTransportPort)
+		return
+	}
+	tm.Logger.Infof("listening on network: %s", netType)
+	tm.listenersMu.Lock()
+	tm.listeners = append(tm.listeners, lis)
+	tm.listenersMu.Unlock()
 
-		if tm.isClosing() {
-			return
-		}
+	if tm.isClosing() {
+		return
+	}
 
-		tm.wgMu.Lock()
-		tm.wg.Add(1)
-		tm.wgMu.Unlock()
+	tm.wgMu.Lock()
+	tm.wg.Add(1)
+	tm.wgMu.Unlock()
 
-		go func() {
-			defer tm.wg.Done()
-			for {
-				select {
-				case <-ctx.Done():
-					return
-				case <-tm.done:
-					return
-				default:
-					if err := tm.acceptTransport(ctx, lis); err != nil {
-						tm.Logger.Warnf("Failed to accept connection: %v", err)
-						if strings.Contains(err.Error(), "closed") {
-							return
-						}
+	go func() {
+		defer tm.wg.Done()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-tm.done:
+				return
+			default:
+				if err := tm.acceptTransport(ctx, lis); err != nil {
+					tm.Logger.Warnf("Failed to accept connection: %v", err)
+					if strings.Contains(err.Error(), "closed") {
+						return
 					}
 				}
 			}
-		}()
+		}
+	}()
+}
+
+func (tm *Manager) serve(ctx context.Context) {
+	// some networks may not be ready yet, so we're setting a callback first
+	tm.n.OnNewNetworkType(func(netType string) {
+		tm.serveNetwork(ctx, netType)
```

Comment:
If we make arclient be responsible for redialing, we don't need this callback logic here.

Instead, we should make `tm.serveNetwork` return an error when the `n.Listen` call fails, and wrap this with some kind of redialing logic.

Reply:
@evanlinjin well, 2 things here. I totally forgot about the error being returned there, that's fair. but I'd stick to the callback just for one reason. if we put redialing logic here, we'll be having AR rediailing and redialing-per-failed-network-type here. which may make start of a visor painful for small machines, for RPi, for example

Reply:
@Darkren I think it is either we completely use this callback, or do the loop through `n.TransportNetworks()`. Having both is kind of confusing in my opinion.

Reply:
@evanlinjin that's fair. I don't really like callbacks that much too. but choosing between callback and beating up CPU I prefered the first one. well, we have a lot of different network types rn. @nkryuchkov was going to generalize code there a bit. probably an interface for network would be good with method like `Ready` which exposes channel. this way we would be able to setup just goroutines for networks like `stcpr/stcph`, wait for a signal on the channel and then start servingn network. that would be much cleaner I think. but this is not really possible rn before generalization is done

Reply:
Yes, it will be good to have something like `Ready() <-chan struct{}` as a standard across all transport types. For the meantime, could you make a note?

I think making this standard is priority, otherwise the codebase will just get messier.

Reply:
@evanlinjin sure, will add `TODO` there